Warn when `javascript:` URLs are used during client-side navigation

eps1lon opened the pull request:
Only includes App Router in this draft. Page Router support will be added in the same PR. Looking for feedback on the naming, `javascript:` detection mechanism, and if we should include `data:` (which also needs confirmation from React because React only guards against `javascript:` not `data:`).

Triggers a dev-only warning when `javascript:` URLs are navigated to during client-side navigation (`router.push(href)`, `router.replace(href)`, `router.prefetch(href)`, `<Link href={href} as={href} />`). To opt-out, pass `{ unsafeHref: { __href: href } }` instead e.g. `router.push({ unsafeHref: { __href: userInput } })`. If the page isn't hydrated yet, React 19 will prevent execution of `javascript:` URLs in anchor tags.
We also introduce an experimental `hardenedXSSProtection` config that will completely prevent execution of untrusted `javascript:` URLs by throwing. In a future release, this behavior will become the default.

[Docs preview](https://github.com/vercel/next.js/blob/sebbie/validate-javascript-uris/docs/02-app/01-building-your-application/07-configuring/15-content-security-policy.mdx)

TODO:
- [ ] Pages Router
- [x] docs

Closes NEXT-3140

eps1lon committed:
Warn when javascript: URLs are used during client-side navigation

eps1lon committed:
Ignore pages router for now

ijjk commented:
## Failing test suites <!--22940400680-->
Commit: eb75f1bcfab88cfb08361aab144be7db8a59cdbf

`__NEXT_EXPERIMENTAL_PPR=true pnpm test-dev test/e2e/app-dir/router-xss/router-xss.test.ts` (PPR)
- default > warns on javascript URLs for client-side navigation with as
- hardenedXSSProtection > warns on javascript URLs for client-side navigation

<details>
<summary>Expand output</summary>

  ● default › warns on javascript URLs for client-side navigation with as

    expect(received).toEqual(expected) // deep equality

    - Expected  - 0
    + Received  + 4

    @@ -1,7 +1,11 @@
      Array [
        Object {
    +     "message": "received ws message {\"action\":\"sync\",\"hash\":\"fe34502f3a9826cb\",\"errors\":[],\"warnings\":[],\"versionInfo\":{\"staleness\":\"fresh\",\"installed\":\"14.3.0-canary.11\"}}",
    +     "source": "log",
    +   },
    +   Object {
          "message": "A future version of Next.js will block `javascript:` URLs as a security precaution. Use event handlers instead if you can. If you need to push unsafe URLs, use `router.push({ unsafeHref: { __href: href } })` instead. A client-side navigation to \"%s\" was triggered. \"javascript:console.log('XSS untrusted client-side navigation with as');\"",
          "source": "error",
        },
        Object {
          "message": "XSS untrusted client-side navigation with as",

      102 |         initialConsoleCalls.length
      103 |       )
    > 104 |       expect(consoleCallsAfterInteraction).toEqual(
          |                                            ^
      105 |         isNextDev
      106 |           ? [
      107 |               {

      at toEqual (e2e/app-dir/router-xss/router-xss.test.ts:104:44)
      at fn (lib/next-test-utils.ts:730:20)
      at Object.<anonymous> (e2e/app-dir/router-xss/router-xss.test.ts:100:5)

  ● hardenedXSSProtection › warns on javascript URLs for client-side navigation

    expect(received).toEqual(expected) // deep equality

    - Expected  - 0
    + Received  + 4

      Array [
        Object {
    +     "message": "received ws message {\"action\":\"sync\",\"hash\":\"18b1977f4c89dbc4\",\"errors\":[],\"warnings\":[],\"versionInfo\":{\"staleness\":\"fresh\",\"installed\":\"14.3.0-canary.11\"}}",
    +     "source": "log",
    +   },
    +   Object {
          "message": "Next.js has blocked a `javascript:` URL as a security precaution.",
          "source": "error",
        },
      ]

      305 |         initialConsoleCalls.length
      306 |       )
    > 307 |       expect(consoleCallsAfterInteraction).toEqual([
          |                                            ^
      308 |         {
      309 |           source: 'error',
      310 |           message:

      at toEqual (e2e/app-dir/router-xss/router-xss.test.ts:307:44)
      at fn (lib/next-test-utils.ts:730:20)
      at Object.<anonymous> (e2e/app-dir/router-xss/router-xss.test.ts:303:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

eps1lon committed:
Split test pages to reduce noise

eps1lon committed:
Add unit tests

eps1lon committed:
Test that the cases are actually JavaScript protocols

eps1lon commented on packages/next/src/server/config-shared.ts:
```diff
@@ -444,6 +444,11 @@ export interface ExperimentalConfig {
   testProxy?: boolean
+
+  /**
+   * Prevents client-side navigation to URLs that contain a `javascript:` schema.
```
```suggestion
   * Prevents client-side navigation to URLs that use the `javascript:` protocol.
```

eps1lon commented on packages/next/src/shared/lib/xss-protection.ts:
```diff
@@ -0,0 +1,46 @@
+  }
+}
+
+// TODO: Should this handle special chars?
```
Doesn't seem like it: https://vercel.slack.com/archives/C01RUTYJYHW/p1713535684819809?thread_ts=1712932117.992689&cid=C01RUTYJYHW

eps1lon commented on packages/next/src/shared/lib/xss-protection.ts:
```diff
@@ -0,0 +1,46 @@
+  /^[\u0000-\u001F ]*j[\r\n\t]*a[\r\n\t]*v[\r\n\t]*a[\r\n\t]*s[\r\n\t]*c[\r\n\t]*r[\r\n\t]*i[\r\n\t]*p[\r\n\t]*t[\r\n\t]*\:/i
+export function isJavaScriptProtocol(href: string) {
+  // We could just use `new URL(href).protocol === 'javascript:'`
+  // But that might be more expensive than this simple regex may not be available everywhere.
```
```suggestion
  // But that might be more expensive than this simple regex since it also throws on invalid URLs and `URL` may not be available in every environemnt.
```

eps1lon committed:
Assert on normalized (turbopack vs non-turbopack) messages

eps1lon committed:
Don't need to handle special chars

See https://yp39gm.csb.app/

ijjk commented:
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/validate-javascript-uris | Change |
| - | - | - | - |
| buildDuration | 15.9s | 15.9s | N/A |
| buildDurationCached | 8.7s | 7.3s | N/A |
| nodeModulesSize | 199 MB | 199 MB | ⚠️ +23.7 kB |
| nextStartRea..uration (ms) | 403ms | 430ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/validate-javascript-uris | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 31.5 kB | 31.6 kB | ⚠️ +137 B |
| 3304.HASH.js gzip | 169 B | 169 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 8299-HASH.js gzip | 5.1 kB | 5.1 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 228 B | 227 B | N/A |
| main-HASH.js gzip | 29.6 kB | 29.7 kB | N/A |
| webpack-HASH.js gzip | 1.64 kB | 1.65 kB | N/A |
| Overall change | 76.9 kB | 77 kB | ⚠️ +137 B |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/validate-javascript-uris | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/validate-javascript-uris | Change |
| - | - | - | - |
| _app-HASH.js gzip | 193 B | 194 B | N/A |
| _error-HASH.js gzip | 193 B | 191 B | N/A |
| amp-HASH.js gzip | 511 B | 511 B | ✓ |
| css-HASH.js gzip | 342 B | 343 B | N/A |
| dynamic-HASH.js gzip | 2.51 kB | 2.51 kB | N/A |
| edge-ssr-HASH.js gzip | 265 B | 265 B | ✓ |
| head-HASH.js gzip | 365 B | 364 B | N/A |
| hooks-HASH.js gzip | 389 B | 391 B | N/A |
| image-HASH.js gzip | 4.28 kB | 4.28 kB | N/A |
| index-HASH.js gzip | 269 B | 268 B | N/A |
| link-HASH.js gzip | 2.68 kB | 2.69 kB | N/A |
| routerDirect..HASH.js gzip | 328 B | 326 B | N/A |
| script-HASH.js gzip | 395 B | 397 B | N/A |
| withRouter-HASH.js gzip | 323 B | 323 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.21 kB | 1.21 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/validate-javascript-uris | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 485 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/validate-javascript-uris | Change |
| - | - | - | - |
| index.html gzip | 529 B | 530 B | N/A |
| link.html gzip | 540 B | 542 B | N/A |
| withRouter.html gzip | 523 B | 523 B | ✓ |
| Overall change | 523 B | 523 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/validate-javascript-uris | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94.5 kB | 94.6 kB | N/A |
| page.js gzip | 3.05 kB | 3.06 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/validate-javascript-uris | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 623 B | 626 B | N/A |
| middleware-r..fest.js gzip | 155 B | 156 B | N/A |
| middleware.js gzip | 25.6 kB | 25.6 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 839 B | 839 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/validate-javascript-uris | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 97.6 kB | 97.6 kB | ✓ |
| app-page-tur..prod.js gzip | 99.4 kB | 99.4 kB | ✓ |
| app-page-tur..prod.js gzip | 93.6 kB | 93.6 kB | ✓ |
| app-page.run...dev.js gzip | 145 kB | 145 kB | ✓ |
| app-page.run..prod.js gzip | 92.1 kB | 92.1 kB | ✓ |
| app-route-ex...dev.js gzip | 21.5 kB | 21.5 kB | ✓ |
| app-route-ex..prod.js gzip | 15.1 kB | 15.1 kB | ✓ |
| app-route-tu..prod.js gzip | 15.1 kB | 15.1 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route.ru...dev.js gzip | 21.2 kB | 21.2 kB | ✓ |
| app-route.ru..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 21.4 kB | 21.5 kB | N/A |
| pages.runtim...dev.js gzip | 22.1 kB | 22.1 kB | N/A |
| pages.runtim..prod.js gzip | 21.4 kB | 21.5 kB | N/A |
| server.runti..prod.js gzip | 51.6 kB | 51.6 kB | ✓ |
| Overall change | 881 kB | 881 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/validate-javascript-uris | Change |
| - | - | - | - |
| 0.pack gzip | 1.6 MB | 1.6 MB | ⚠️ +1.27 kB |
| index.pack gzip | 107 kB | 108 kB | ⚠️ +873 B |
| Overall change | 1.7 MB | 1.71 MB | ⚠️ +2.14 kB |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>image-HASH.js</strong></summary>

```diff
@@ -1,7 +1,7 @@
 (self["webpackChunk_N_E"] = self["webpackChunk_N_E"] || []).push([
   [8358],
   {
-    /***/ 1552: /***/ (
+    /***/ 4070: /***/ (
       __unused_webpack_module,
       __unused_webpack_exports,
       __webpack_require__
@@ -9,7 +9,7 @@
       (window.__NEXT_P = window.__NEXT_P || []).push([
         "/image",
         function () {
-          return __webpack_require__(5237);
+          return __webpack_require__(396);
         },
       ]);
       if (false) {
@@ -18,7 +18,7 @@
       /***/
     },
 
-    /***/ 2016: /***/ (module, exports, __webpack_require__) => {
+    /***/ 8490: /***/ (module, exports, __webpack_require__) => {
       "use strict";
       /* __next_internal_client_entry_do_not_use__  cjs */
       Object.defineProperty(exports, "__esModule", {
@@ -40,15 +40,15 @@
         __webpack_require__(422)
       );
       const _head = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(6074)
+        __webpack_require__(2457)
       );
-      const _getimgprops = __webpack_require__(9571);
-      const _imageconfig = __webpack_require__(6567);
-      const _imageconfigcontextsharedruntime = __webpack_require__(419);
-      const _warnonce = __webpack_require__(4486);
-      const _routercontextsharedruntime = __webpack_require__(162);
+      const _getimgprops = __webpack_require__(7932);
+      const _imageconfig = __webpack_require__(5706);
+      const _imageconfigcontextsharedruntime = __webpack_require__(9483);
+      const _warnonce = __webpack_require__(9035);
+      const _routercontextsharedruntime = __webpack_require__(4829);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(6996)
+        __webpack_require__(7240)
       );
       // This is replaced by webpack define plugin
       const configEnv = {
@@ -379,7 +379,7 @@
       /***/
     },
 
-    /***/ 9571: /***/ (
+    /***/ 7932: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -395,9 +395,9 @@
           return getImgProps;
         },
       });
-      const _warnonce = __webpack_require__(4486);
-      const _imageblursvg = __webpack_require__(133);
-      const _imageconfig = __webpack_require__(6567);
+      const _warnonce = __webpack_require__(9035);
+      const _imageblursvg = __webpack_require__(2642);
+      const _imageconfig = __webpack_require__(5706);
       const VALID_LOADING_VALUES =
         /* unused pure expression or super */ null && [
           "lazy",
@@ -772,7 +772,7 @@
       /***/
     },
 
-    /***/ 133: /***/ (__unused_webpack_module, exports) => {
+    /***/ 2642: /***/ (__unused_webpack_module, exports) => {
       "use strict";
       /**
        * A shared function, used on both client and server, to generate a SVG blur placeholder.
@@ -827,7 +827,7 @@
       /***/
     },
 
-    /***/ 4085: /***/ (
+    /***/ 503: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -854,10 +854,10 @@
         },
       });
       const _interop_require_default = __webpack_require__(2430);
-      const _getimgprops = __webpack_require__(9571);
-      const _imagecomponent = __webpack_require__(2016);
+      const _getimgprops = __webpack_require__(7932);
+      const _imagecomponent = __webpack_require__(8490);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(6996)
+        __webpack_require__(7240)
       );
       function getImageProps(imgProps) {
         const { props } = (0, _getimgprops.getImgProps)(imgProps, {
@@ -889,7 +889,7 @@
       /***/
     },
 
-    /***/ 6996: /***/ (__unused_webpack_module, exports) => {
+    /***/ 7240: /***/ (__unused_webpack_module, exports) => {
       "use strict";
 
       Object.defineProperty(exports, "__esModule", {
@@ -924,7 +924,7 @@
       /***/
     },
 
-    /***/ 5237: /***/ (
+    /***/ 396: /***/ (
       __unused_webpack_module,
       __webpack_exports__,
       __webpack_require__
@@ -941,8 +941,8 @@
 
       // EXTERNAL MODULE: ./node_modules/.pnpm/react@18.2.0/node_modules/react/jsx-runtime.js
       var jsx_runtime = __webpack_require__(1527);
-      // EXTERNAL MODULE: ./node_modules/.pnpm/file+..+main-repo+packages+next+next-packed.tgz_react-dom@18.2.0_react@18.2.0/node_modules/next/image.js
-      var next_image = __webpack_require__(1577);
+      // EXTERNAL MODULE: ./node_modules/.pnpm/file+..+diff-repo+packages+next+next-packed.tgz_react-dom@18.2.0_react@18.2.0/node_modules/next/image.js
+      var next_image = __webpack_require__(73);
       var image_default = /*#__PURE__*/ __webpack_require__.n(next_image); // CONCATENATED MODULE: ./pages/nextjs.png
       /* harmony default export */ const nextjs = {
         src: "/_next/static/media/nextjs.cae0b805.png",
@@ -972,12 +972,8 @@
       /***/
     },
 
-    /***/ 1577: /***/ (
-      module,
-      __unused_webpack_exports,
-      __webpack_require__
-    ) => {
-      module.exports = __webpack_require__(4085);
+    /***/ 73: /***/ (module, __unused_webpack_exports, __webpack_require__) => {
+      module.exports = __webpack_require__(503);
 
       /***/
     },
@@ -987,7 +983,7 @@
     /******/ var __webpack_exec__ = (moduleId) =>
       __webpack_require__((__webpack_require__.s = moduleId));
     /******/ __webpack_require__.O(0, [2888, 9774, 179], () =>
-      __webpack_exec__(1552)
+      __webpack_exec__(4070)
     );
     /******/ var __webpack_exports__ = __webpack_require__.O();
     /******/ _N_E = __webpack_exports__;
```
</details>
<details>
<summary>Diff for <strong>2453-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages-turbo...time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages.runtime.dev.js</strong></summary>

```diff
@@ -14,7 +14,7 @@ Read more: https://nextjs.org/docs/messages/${r}`};function eY(e,t,r){let{destin
 `+i.join(" and ")+"\nSee more info here: https://nextjs.org/docs/messages/invalid-redirect-gssp")}async function eK(n,s,a,i,l,d){var u,c,p;let h,m,f,g;(0,x.gk)({req:n},"cookies",(c=n.headers,function(){let{cookie:e}=c;if(!e)return{};let{parse:t}=r("./dist/compiled/cookie/index.js");return t(Array.isArray(e)?e.join("; "):e)}));let v={};if(v.assetQueryString=l.dev&&l.assetQueryString||"",l.dev&&!v.assetQueryString){let e=(n.headers["user-agent"]||"").toLowerCase();e.includes("safari")&&!e.includes("chrome")&&(v.assetQueryString=`?ts=${Date.now()}`)}l.deploymentId&&(v.assetQueryString+=`${v.assetQueryString?"&":"?"}dpl=${l.deploymentId}`),i=Object.assign({},i);let{err:b,dev:w=!1,ampPath:S="",pageConfig:R={},buildManifest:E,reactLoadableManifest:T,ErrorDebug:$,getStaticProps:I,getStaticPaths:k,getServerSideProps:M,isDataReq:D,params:F,previewProps:H,basePath:W,images:Y,runtime:K,isExperimentalCompile:et,swrDelta:es}=l,{App:ea}=d,ed=v.assetQueryString,eu=d.Document,ec=l.Component,ep=!!i.__nextFallback,eh=i.__nextNotFoundSrcPage;!function(e){for(let t of e$)delete e[t]}(i);let em=!!I,ef=em&&l.nextExport,eg=ea.getInitialProps===ea.origGetInitialProps,ev=!!(null==ec?void 0:ec.getInitialProps),ey=null==ec?void 0:ec.unstable_scriptLoader,ex=G(a),eb="/_error"===a&&ec.getInitialProps===ec.origGetInitialProps;l.nextExport&&ev&&!eb&&t(`Detected getInitialProps on page '${a}' while running export. It's recommended to use getStaticProps which has a more correct behavior for static exporting.
 Read more: https://nextjs.org/docs/messages/get-initial-props-export`);let ew=!ev&&eg&&!em&&!M;if(ew&&!w&&et&&(s.setHeader("Cache-Control",function({revalidate:e,swrDelta:t}){let r=t?`stale-while-revalidate=${t}`:"stale-while-revalidate";return 0===e?"private, no-cache, no-store, max-age=0, must-revalidate":"number"==typeof e?`s-maxage=${e}, ${r}`:`s-maxage=${C.BR}, ${r}`}({revalidate:!1,swrDelta:es})),ew=!1),ev&&em)throw Error(C.wh+` ${a}`);if(ev&&M)throw Error(C.Wo+` ${a}`);if(M&&em)throw Error(C.oL+` ${a}`);if(M&&"export"===l.nextConfigOutput)throw Error('getServerSideProps cannot be used with "output: export". See more info here: https://nextjs.org/docs/advanced-features/static-html-export');if(k&&!ex)throw Error(`getStaticPaths is only allowed for dynamic SSG pages and was found on '${a}'.
 Read more: https://nextjs.org/docs/messages/non-dynamic-getstaticpaths-usage`);if(k&&!em)throw Error(`getStaticPaths was added without a getStaticProps in ${a}. Without getStaticProps, getStaticPaths does nothing`);if(em&&ex&&!k)throw Error(`getStaticPaths is required for dynamic SSG pages and is missing for '${a}'.
-Read more: https://nextjs.org/docs/messages/invalid-getstaticpaths-value`);let eS=l.resolvedAsPath||n.url;if(w){let{isValidElementType:e}=r("./dist/compiled/react-is/index.js");if(!e(ec))throw Error(`The default export is not a React Component in page: "${a}"`);if(!e(ea))throw Error('The default export is not a React Component in page: "/_app"');if(!e(eu))throw Error('The default export is not a React Component in page: "/_document"');if((ew||ep)&&(i={...i.amp?{amp:i.amp}:{}},eS=`${a}${n.url.endsWith("/")&&"/"!==a&&!ex?"/":""}`,n.url=a),"/404"===a&&(ev||M))throw Error(`\`pages/404\` ${C.Ei}`);if(j.Er.includes(a)&&(ev||M))throw Error(`\`pages${a}\` ${C.Ei}`)}for(let e of["getStaticProps","getServerSideProps","getStaticPaths"])if(null==ec?void 0:ec[e])throw Error(`page ${a} ${e} ${C.lk}`);await q.preloadAll(),(em||M)&&!ep&&H&&(f=!1!==(h=e(n,s,H)));let eP=new eX(a,i,eS,{isFallback:ep},!!(M||ev||!eg&&!em||et),W,l.locale,l.locales,l.defaultLocale,l.domainLocales,f,ee(n,"isLocaleDomain")),e_={back(){eP.back()},forward(){eP.forward()},refresh(){eP.reload()},fastRefresh(){},push(e,t){let{scroll:r}=void 0===t?{}:t;eP.push(e,void 0,{scroll:r})},replace(e,t){let{scroll:r}=void 0===t?{}:t;eP.replace(e,void 0,{scroll:r})},prefetch(e){eP.prefetch(e)}},eC={},eE=(0,_.createStyleRegistry)(),eA={ampFirst:!0===R.amp,hasQuery:!!i.amp,hybrid:"hybrid"===R.amp},eO=function(e){let{ampFirst:t=!1,hybrid:r=!1,hasQuery:n=!1}=void 0===e?{}:e;return t||r&&n}(eA),eI=function(e){void 0===e&&(e=!1);let t=[(0,y.jsx)("meta",{charSet:"utf-8"})];return e||t.push((0,y.jsx)("meta",{name:"viewport",content:"width=device-width"})),t}(eO),eH=[],eq={};ey&&(eq.beforeInteractive=[].concat(ey()).filter(e=>"beforeInteractive"===e.props.strategy).map(e=>e.props));let ez=({children:e})=>(0,y.jsx)(eF.Provider,{value:e_,children:(0,y.jsx)(eN.Provider,{value:eP.isReady&&eP.query?new URL(eP.asPath,"http://n").searchParams:new URLSearchParams,children:(0,y.jsx)(eD,{router:eP,isAutoExport:ew,children:(0,y.jsx)(eL.Provider,{value:function(e){if(!e.isReady||!e.query)return null;let t={};for(let r of Object.keys(function(e){let{parameterizedRoute:t,groups:r}=function(e){let t=el(e).slice(1).split("/"),r={},n=1;return{parameterizedRoute:t.map(e=>{let t=U.find(t=>e.startsWith(t)),o=e.match(/\[((?:\[.*\])|.+)\]/);if(t&&o){let{key:e,optional:s,repeat:a}=eM(o[1]);return r[e]={pos:n++,repeat:a,optional:s},"/"+ek(t)+"([^/]+?)"}if(!o)return"/"+ek(e);{let{key:e,repeat:t,optional:s}=eM(o[1]);return r[e]={pos:n++,repeat:t,optional:s},t?s?"(?:/(.+?))?":"/(.+?)":"/([^/]+?)"}}).join(""),groups:r}}(e);return{re:RegExp("^"+t+"(?:/)?$"),groups:r}}(e.pathname).groups))t[r]=e.query[r];return t}(eP),children:(0,y.jsx)(z.Provider,{value:eP,children:(0,y.jsx)(A.Provider,{value:eA,children:(0,y.jsx)(L.Provider,{value:{updateHead:e=>{eI=e},updateScripts:e=>{eC=e},scripts:eq,mountedInstances:new Set},children:(0,y.jsx)(O.Provider,{value:e=>eH.push(e),children:(0,y.jsx)(_.StyleRegistry,{registry:eE,children:(0,y.jsx)(eR.Provider,{value:Y,children:e})})})})})})})})})}),eB=()=>null,eJ=({children:e})=>(0,y.jsxs)(y.Fragment,{children:[(0,y.jsx)(eB,{}),(0,y.jsx)(ez,{children:(0,y.jsxs)(y.Fragment,{children:[w?(0,y.jsxs)(y.Fragment,{children:[e,(0,y.jsx)(eB,{})]}):e,(0,y.jsx)(eB,{})]})})]}),eK={err:b,req:ew?void 0:n,res:ew?void 0:s,pathname:a,query:i,asPath:eS,locale:l.locale,locales:l.locales,defaultLocale:l.defaultLocale,AppTree:e=>(0,y.jsx)(eJ,{children:eQ(ea,ec,{...e,router:eP})}),defaultGetInitialProps:async(e,t={})=>{let{html:r,head:n}=await e.renderPage({enhanceApp:e=>t=>(0,y.jsx)(e,{...t})}),o=eE.styles({nonce:t.nonce});return eE.flush(),{html:r,head:n,styles:o}}},e0=!em&&(l.nextExport||w&&(ew||ep)),e1=()=>{let e=eE.styles();return eE.flush(),(0,y.jsx)(y.Fragment,{children:e})};if(m=await X(ea,{AppTree:eK.AppTree,Component:ec,router:eP,ctx:eK}),(em||M)&&f&&(m.__N_PREVIEW=!0),em&&(m[j.NO]=!0),em&&!ep){let e,t;try{e=await (0,en.getTracer)().trace(eo.xj.getStaticProps,{spanName:`getStaticProps ${a}`,attributes:{"next.route":a}},()=>I({...ex?{params:i}:void 0,...f?{draftMode:!0,preview:!0,previewData:h}:void 0,locales:l.locales,locale:l.locale,defaultLocale:l.defaultLocale,revalidateReason:l.isOnDemandRevalidate?"on-demand":ef?"build":"stale"}))}catch(e){throw e&&"ENOENT"===e.code&&delete e.code,e}if(null==e)throw Error(C.q6);let r=Object.keys(e).filter(e=>"revalidate"!==e&&"props"!==e&&"redirect"!==e&&"notFound"!==e);if(r.includes("unstable_revalidate"))throw Error(C.Eo);if(r.length)throw Error(eZ("getStaticProps",r));if(void 0!==e.notFound&&void 0!==e.redirect)throw Error(`\`redirect\` and \`notFound\` can not both be returned from ${em?"getStaticProps":"getServerSideProps"} at the same time. Page: ${a}
+Read more: https://nextjs.org/docs/messages/invalid-getstaticpaths-value`);let eS=l.resolvedAsPath||n.url;if(w){let{isValidElementType:e}=r("./dist/compiled/react-is/index.js");if(!e(ec))throw Error(`The default export is not a React Component in page: "${a}"`);if(!e(ea))throw Error('The default export is not a React Component in page: "/_app"');if(!e(eu))throw Error('The default export is not a React Component in page: "/_document"');if((ew||ep)&&(i={...i.amp?{amp:i.amp}:{}},eS=`${a}${n.url.endsWith("/")&&"/"!==a&&!ex?"/":""}`,n.url=a),"/404"===a&&(ev||M))throw Error(`\`pages/404\` ${C.Ei}`);if(j.Er.includes(a)&&(ev||M))throw Error(`\`pages${a}\` ${C.Ei}`)}for(let e of["getStaticProps","getServerSideProps","getStaticPaths"])if(null==ec?void 0:ec[e])throw Error(`page ${a} ${e} ${C.lk}`);await q.preloadAll(),(em||M)&&!ep&&H&&(f=!1!==(h=e(n,s,H)));let eP=new eX(a,i,eS,{isFallback:ep},!!(M||ev||!eg&&!em||et),W,l.locale,l.locales,l.defaultLocale,l.domainLocales,f,ee(n,"isLocaleDomain")),e_={back(){eP.back()},forward(){eP.forward()},refresh(){eP.reload()},fastRefresh(){},push(e,t){let{scroll:r}=void 0===t?{}:t;eP.push("string"==typeof e?e:e.unsafeHref.__href,void 0,{scroll:r})},replace(e,t){let{scroll:r}=void 0===t?{}:t;eP.replace("string"==typeof e?e:e.unsafeHref.__href,void 0,{scroll:r})},prefetch(e){eP.prefetch("string"==typeof e?e:e.unsafeHref.__href)}},eC={},eE=(0,_.createStyleRegistry)(),eA={ampFirst:!0===R.amp,hasQuery:!!i.amp,hybrid:"hybrid"===R.amp},eO=function(e){let{ampFirst:t=!1,hybrid:r=!1,hasQuery:n=!1}=void 0===e?{}:e;return t||r&&n}(eA),eI=function(e){void 0===e&&(e=!1);let t=[(0,y.jsx)("meta",{charSet:"utf-8"})];return e||t.push((0,y.jsx)("meta",{name:"viewport",content:"width=device-width"})),t}(eO),eH=[],eq={};ey&&(eq.beforeInteractive=[].concat(ey()).filter(e=>"beforeInteractive"===e.props.strategy).map(e=>e.props));let ez=({children:e})=>(0,y.jsx)(eF.Provider,{value:e_,children:(0,y.jsx)(eN.Provider,{value:eP.isReady&&eP.query?new URL(eP.asPath,"http://n").searchParams:new URLSearchParams,children:(0,y.jsx)(eD,{router:eP,isAutoExport:ew,children:(0,y.jsx)(eL.Provider,{value:function(e){if(!e.isReady||!e.query)return null;let t={};for(let r of Object.keys(function(e){let{parameterizedRoute:t,groups:r}=function(e){let t=el(e).slice(1).split("/"),r={},n=1;return{parameterizedRoute:t.map(e=>{let t=U.find(t=>e.startsWith(t)),o=e.match(/\[((?:\[.*\])|.+)\]/);if(t&&o){let{key:e,optional:s,repeat:a}=eM(o[1]);return r[e]={pos:n++,repeat:a,optional:s},"/"+ek(t)+"([^/]+?)"}if(!o)return"/"+ek(e);{let{key:e,repeat:t,optional:s}=eM(o[1]);return r[e]={pos:n++,repeat:t,optional:s},t?s?"(?:/(.+?))?":"/(.+?)":"/([^/]+?)"}}).join(""),groups:r}}(e);return{re:RegExp("^"+t+"(?:/)?$"),groups:r}}(e.pathname).groups))t[r]=e.query[r];return t}(eP),children:(0,y.jsx)(z.Provider,{value:eP,children:(0,y.jsx)(A.Provider,{value:eA,children:(0,y.jsx)(L.Provider,{value:{updateHead:e=>{eI=e},updateScripts:e=>{eC=e},scripts:eq,mountedInstances:new Set},children:(0,y.jsx)(O.Provider,{value:e=>eH.push(e),children:(0,y.jsx)(_.StyleRegistry,{registry:eE,children:(0,y.jsx)(eR.Provider,{value:Y,children:e})})})})})})})})})}),eB=()=>null,eJ=({children:e})=>(0,y.jsxs)(y.Fragment,{children:[(0,y.jsx)(eB,{}),(0,y.jsx)(ez,{children:(0,y.jsxs)(y.Fragment,{children:[w?(0,y.jsxs)(y.Fragment,{children:[e,(0,y.jsx)(eB,{})]}):e,(0,y.jsx)(eB,{})]})})]}),eK={err:b,req:ew?void 0:n,res:ew?void 0:s,pathname:a,query:i,asPath:eS,locale:l.locale,locales:l.locales,defaultLocale:l.defaultLocale,AppTree:e=>(0,y.jsx)(eJ,{children:eQ(ea,ec,{...e,router:eP})}),defaultGetInitialProps:async(e,t={})=>{let{html:r,head:n}=await e.renderPage({enhanceApp:e=>t=>(0,y.jsx)(e,{...t})}),o=eE.styles({nonce:t.nonce});return eE.flush(),{html:r,head:n,styles:o}}},e0=!em&&(l.nextExport||w&&(ew||ep)),e1=()=>{let e=eE.styles();return eE.flush(),(0,y.jsx)(y.Fragment,{children:e})};if(m=await X(ea,{AppTree:eK.AppTree,Component:ec,router:eP,ctx:eK}),(em||M)&&f&&(m.__N_PREVIEW=!0),em&&(m[j.NO]=!0),em&&!ep){let e,t;try{e=await (0,en.getTracer)().trace(eo.xj.getStaticProps,{spanName:`getStaticProps ${a}`,attributes:{"next.route":a}},()=>I({...ex?{params:i}:void 0,...f?{draftMode:!0,preview:!0,previewData:h}:void 0,locales:l.locales,locale:l.locale,defaultLocale:l.defaultLocale,revalidateReason:l.isOnDemandRevalidate?"on-demand":ef?"build":"stale"}))}catch(e){throw e&&"ENOENT"===e.code&&delete e.code,e}if(null==e)throw Error(C.q6);let r=Object.keys(e).filter(e=>"revalidate"!==e&&"props"!==e&&"redirect"!==e&&"notFound"!==e);if(r.includes("unstable_revalidate"))throw Error(C.Eo);if(r.length)throw Error(eZ("getStaticProps",r));if(void 0!==e.notFound&&void 0!==e.redirect)throw Error(`\`redirect\` and \`notFound\` can not both be returned from ${em?"getStaticProps":"getServerSideProps"} at the same time. Page: ${a}
 See more info here: https://nextjs.org/docs/messages/gssp-mixed-not-found-redirect`);if("notFound"in e&&e.notFound){if("/404"===a)throw Error('The /404 page can not return notFound in "getStaticProps", please remove it to continue!');v.isNotFound=!0}if("redirect"in e&&e.redirect&&"object"==typeof e.redirect){if(eY(e.redirect,n,"getStaticProps"),ef)throw Error(`\`redirect\` can not be returned from getStaticProps during prerendering (${n.url})
 See more info here: https://nextjs.org/docs/messages/gsp-redirect-during-prerender`);e.props={__N_REDIRECT:e.redirect.destination,__N_REDIRECT_STATUS:er(e.redirect)},void 0!==e.redirect.basePath&&(e.props.__N_REDIRECT_BASE_PATH=e.redirect.basePath),v.isRedirect=!0}if((w||ef)&&!v.isNotFound&&!N(a,"getStaticProps",e.props))throw Error("invariant: getStaticProps did not return valid props. Please report this.");if("revalidate"in e){if(e.revalidate&&"export"===l.nextConfigOutput)throw Error('ISR cannot be used with "output: export". See more info here: https://nextjs.org/docs/advanced-features/static-html-export');if("number"==typeof e.revalidate){if(Number.isInteger(e.revalidate)){if(e.revalidate<=0)throw Error(`A page's revalidate option can not be less than or equal to zero for ${n.url}. A revalidate option of zero means to revalidate after _every_ request, and implies stale data cannot be tolerated.
```
</details>
<details>
<summary>Diff for <strong>pages.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 5333d22b09b9cf9f2937f4678ac32eb91e718f42

eps1lon committed:
Docs

eps1lon committed:
lint-no-typescript does not support nested tsconfigs

eps1lon committed:
Ignore logs from fast refresh

eps1lon committed:
Flake hardening

eps1lon committed:
Test cleanup

timneutkens commented on docs/02-app/01-building-your-application/07-configuring/15-content-security-policy.mdx:
````diff
@@ -242,6 +242,27 @@ module.exports = {
+
+```js filename="next.config.js"
+module.exports = {
+  experimental: {
````
In the past we've used `future` for flags that will be enabled in a future major.
```suggestion
  future: {
```

eps1lon replied:
Done: 6513b89676eb96b4bb0b565a95b03e4bb774e52e

timneutkens commented on packages/next/src/client/components/app-router.tsx:
```diff
@@ -376,14 +379,18 @@ function Router({
       },
-      replace: (href, options = {}) => {
+      replace: (untrustedHref, options = {}) => {
+        // Moving this check into the transition would trigger error boundaries
```
You can call `router.push()` / `router.replace()` / `router.prefetch()` inside of a React transition, not sure it's intended that it would throw in that case?

I.e. this is valid:

```js
startTransition(() => {
  router.push('/')
  router.push('/dashboard')
})
```

Maybe it should `console.error` + ignore the navigation?

eps1lon replied:
True. `console.error` + ignore sounds like a good compromise. Doesn't surface as well though. I'll ask Seb since he did the original implementation in React if he has opinions about it.

eps1lon replied:
Done

timneutkens commented on packages/next/src/shared/lib/router/adapters.tsx:
```diff
@@ -24,13 +24,26 @@ export function adaptForAppRouterInstance(
-      void pagesRouter.push(href, undefined, { scroll })
+      void pagesRouter.push(
+        // TODO: Handle untrusted hrefs
+        typeof href === 'string' ? href : href.unsafeHref.__href,
```
This needs validation in development to show warnings

eps1lon replied:
I was under the impression that this is the code responsible for the Pages router? I planned on doing the implementation in this PR but wanted to make sure the overall behavior makes sense first.

eps1lon committed:
Rewrite the url to a thrower instead

That way we don't trigger error boundaries.
This is how it works in React as well.

eps1lon committed:
Test unicode escapes

eps1lon committed:
Add types for `config.future`

eps1lon committed:
Move from experimental flag to future flag

ztanner commented:
Superseded by https://github.com/vercel/next.js/pull/88185
